### ISSUE TITLE
Consolidate lsm6ds33, LSM6DSL and lsm6dsOX into the same crate?

### ISSUE THREAD

**Issue by gauteh** [nnarain/lsm6ds33#2]
Hi @nnarain  and @Kriskras99 ,

I have just started looking to interface the LSM6DSOX to rust. I noticed this repository as well as https://github.com/Kriskras99/LSM6DSL. There seems to be a lot of overlap between these three devices, perhaps we should try to join efforts and create a LSM6DS driver that supports all three devices. It seems that Adafruit has done so with very little device specific code:

https://github.com/adafruit/Adafruit_LSM6DS

**Comment by nnarain** [comment]
Ya I've been referencing the adafruit drivers when working on this rust implementation.

Definitively not opposed to a consolidated effort. Though I'm not familiar with the other devices (will have to look at the datasheet) and only have access to the LSM6DS33 on my nRF52840 dev board.

**Comment by gauteh** [comment]
I only have the LSM6DSOX:

* https://www.adafruit.com/product/4565

but will also soon start using: https://www.adafruit.com/product/4569 (ISM330DHCX)

which seems to have also a very similar interface (it's in the LSM6DS adafruit library).

**Comment by gauteh** [comment]
In that case it would maybe be more appropriate to name the crate something like `STIM-rs`:

![en inemo_sc1448](https://user-images.githubusercontent.com/56827/143403504-f4a783db-2bd6-48cd-b542-d2d3a28bcaff.jpg)

https://www.st.com/en/mems-and-sensors/inemo-inertial-modules.html

**Comment by Kriskras99** [comment]
I'm all for consolidating our efforts and it does seem they are all basically the same. I'd suggest using `LSM6DS` as the crate name, as STIM implies it supports any and all current and future IMU's by ST.
Any suggestions on how to architecturely implement this?

**Comment by gauteh** [comment]
Yeah, I agree, that also follows adafruit and your existing crates.

tor. 25. nov. 2021, 12:42 skrev Kriskras99 ***@***.***>:

**Comment by gauteh** [comment]
My first idea for implementation was something like:

* One struct per device
* A `IMU/LS6DS` trait implements all the common functionality through default implementations. All we need is:
   * Address
   * write_read(..), maybe even address should be hidden here.
* A configuration trait for each device is probably needed, but can maybe have default impls for the common configuration options.
* the accelerometer and gyroscope traits can probably be generically implemented for anything that has the `IMU/LS6DS` trait.

https://crates.io/crates/accelerometer

Also just discovered this crate: https://crates.io/crates/ism330dhcx

Which means that all devices are pretty much implemented already... don't know if that changes the level of motivation a bit? Would at least make sense to try to get that driver on board as well.

**Comment by sousandrei** [comment]
Coming from the other crate here :eyes: 

Seems like an interesting idea, do you have a draft or something I can get on?

**Comment by Kriskras99** [comment]
Not yet, and I won't be able to create one for at least another week (very busy atm). But you can create one if you want to take a first stab at it
